Fix CI build - and possibility for deadlocks!

ma499 opened the pull request:
CI build should now work. Unit tests no longer deadlock - and deadlock possibility in main library eliminated.

ma499 committed:
Enable builds on CI-debug branch and RDP access to build worker to investigate hanging builds. Build release configuration.

ma499 committed:
Keep RDP session to build worker open on ci-debug branches.

ma499 committed:
Second attempt to keep RDP console to build worker open after build finishes in ci-debug branch.

ma499 committed:
Fix YAML formatting to be valid.

ma499 committed:
Eliminate async deadlocks so that unit tests now pass with maxthreads = 1 (simulating configuration of build worker). Update to latest version of Fluent Assertions. Resolve all Resharper warnings in touched files.

ma499 committed:
Enable parallel builds.

ma499 committed:
Add build badge to Readme.MD and link to Nimbus.

ma499 committed:
Consolidate RDP access scripts. RDP access to build worker now only on ci-debug branch.

ma499 committed:
Revert "Consolidate RDP access scripts. RDP access to build worker now only on ci-debug branch."

This reverts commit d6045d2b21cc7e2b9b635c3de3cabdb7b910cf23.

ma499 committed:
Attempt to stop Powershell error when building pull request to master.

GiampaLab commented:
It's ok for me!

slawomir-brzezinski-at-interxion commented on Aqovia.Utilities.SagaMachine.Tests/StateMachineTests.cs:
```diff
@@ -92,33 +91,23 @@ public StateMachineTests()
-
-
-
+            act.ShouldThrow<SagaException>();
```
`act.ShouldThrow<SagaException>()` is not asyc aware, so it uses `Wait()`, and that's the only reason you need `ConfigureAwait(false)`.
Why not use the newly introduced xUnit 2.0's `Assert.ThrowsAsync` and have it done [the recommended way](https://blog.ciber.no/2014/05/19/using-task-configureawaitfalse-to-prevent-deadlocks-in-async-code/) with just await on it and the test being `async`?

It seems that FluentAssertions sadly didn't catch up yet with the `async` keyword, so if you want a nice `act.ShouldThrowAsync<SagaException>()` extension for the blessed way, you need to roll your own:

```
    public static class FluentAssertionMissingFeaturesExtensions
    {
        public static async Task ThrowsAsync<T>(this Func<Task> testCode) where T : Exception
        {
            await Assert.ThrowsAsync<T>(testCode);
        }
    }
```

ma499 replied:
Ah! Well I didn't realise that when I was figuring through it late night :-) Thanks

However, at least one of the tests uncovered a legitimate deadlock in the library itself, so I may need to add another test explicitly to deal with that.

slawomir-brzezinski-at-interxion replied:
Cool. So you're gonna add that test?
Anyways, good spot, mine is just a cosmetic rant.

ma499 replied:
I struggled with adding a suitable test that doesn't rely on setting maxthreads=1 in xunit.net config. (I tried using ThreadPool.SetMaxThreads to simulate but then I couldn't get it to pass!)

With regards to the Async assert, I tried re-writing the xunit.net way but found it less expressive. I'm thinking of using it as a trigger to contribute an improvement to FluentAssertions rather than writing a custom extension method!

slawomir-brzezinski-at-interxion replied:
Well, not trying to be nasty, as I already said this is cosmetic, but writing the very extension and keeping it in your own code for now can be the first step to proposing it to FluentAssertions.

At least your userspace code, in all the places this is the issue will already be doing things properly.
Anyways, this shouldn't be blocking the release, so really Looks More Than Good Enough for me.

taliesins commented on Aqovia.Utilities.SagaMachine/SagaMachine.cs:
```diff
@@ -30,7 +30,7 @@ public SagaMachine(IKeyValueStore keyValueStore, Func<IEnumerable<ISagaMessageId
         public async Task Handle<TMessage>(TMessage message) where TMessage : ISagaMessageIdentifier
         {
-            await Task.Run(() => ProcessMessage(message));
+            await Task.Run(() => ProcessMessage(message)).ConfigureAwait(false);
```
I think its a problem with the caller of the async code. I wouldn't do it here.

ma499 replied:
But this is a library, so it should ensure that callers can't do something stupid. [Recommendation from Stephen Cleary](http://blog.stephencleary.com/2012/07/dont-block-on-async-code.html#preventing-the-deadlock) (MVP who wrote the [MSDN Async programming guide](https://msdn.microsoft.com/en-us/magazine/jj991977.aspx)) is to set ConfigureAwait(false) on "library" functions whenever possible.

slawomir-brzezinski-at-interxion replied:
I must say that I also found it initially pretty stupid for .NET to introduce a nice single `await` keyword, only to tell everyone that 'the recommended way to use it' is with a boilerplate `.ConfigureAwait(false)`.

After reading [this article](http://blogs.msdn.com/b/pfxteam/archive/2012/06/15/executioncontext-vs-synchronizationcontext.aspx) i now at least understand that this is not the `await` keyword's fault, but rather a design flaw of .NET (read near _'Isn’t SynchronizationContext part of ExecutionContext?'_).

It's just too bad that instead of allowing this design flaw to be fixed in a single place for those who don't depend on the bad design's behaviour using something like `[assembly:ShouldFlowSynchronizationContextByDefault(false)]`, they resort to recommend adorning every use of `await` with some boilerplate.

So, all in all, if it's better to do it for libraries, and it's the best way to do it, we should do it.

RavindraSolanki commented:
It require much more work to than just CI troubleshooting activity.

Commit:
Fixed async deadlock issue

Made "Execute" method async and publish messages with an await that
don't need to run on the captured "SynchronizationContext". This will
prevent deadlock when run in ASP.NET or GUI application.

Commit:
Wrapping async method not necessary in task

Commit:
ExecuteAsync set to private

ExecuteAsync set to private because else we'll have to amend
"SagaMachine.ProcessMessage<TMessage>" to handle async function call
more gracefully.

ma499 commented:
Okay, I think this is good to go now.
